Fixes #33691 - Make preview work for webhooks

ofedoren opened the pull request:
~Needs https://github.com/theforeman/foreman/pull/8637.~

ares commented:
:-( I understand we can't easily add this, but I wish I could pass arbitrary object for the preview, like I'd choose a class and then picked a record from the select drop down, similarly to how we do that for selecting host for other templates. Or we'd pass some mock object as `@object` in the preview mode, so that if I call

`@object.name` would print `"object name"` as a result or `@object.name.upcase` would print `"object name upcase"`.

ofedoren commented:
> :-( I understand we can't easily add this, but I wish I could pass arbitrary object for the preview, like I'd choose a class and then picked a record from the select drop down, similarly to how we do that for selecting host for other templates. Or we'd pass some mock object as `@object` in the preview mode, so that if I call
> 
> `@object.name` would print `"object name"` as a result or `@object.name.upcase` would print `"object name upcase"`.

Hm.. I'll see what I can do, but in case if I don't come up with something now we can return this later. The current issue I see for the first approach is that we'd probably need to change or add something to the exisitng .erb/editor, so it will have more changes in the Foreman (not a big problem, but if we want to cherry-pick this then it might be a problem). With the second approach we would need to deal with the missing methods, since it's not possible to determine what object should be used, which then leads to the first approach which allows users to specify a class to be used to create an `@object` mock.

timogoebel commented:
For the preview, can we use some predefined mock data? How does gitlab do it?

lzap commented:
I considered passing some mock the first moment I saw the Bugzilla, but IMHO this will be more confusing than the current proposal - disabling this feature altogether. For provisioning templates it makes sense because you actually have a reasonable data - a host. But in this case we don't, we would need to mock subscription event as well as some payload. And that will depend on what was picked in the event dropdown, it can be a host, content view, anything. That is a thin ice.

It is my belief that we should write either decent features or skip features altogether.

ares commented:
I'm fine with merging this at least to make it better than it is now. But I still think the mock that just responds to any method and returns a new mock with the name of called method stored for later use would be useful. We may play further with methods like `each`. 

Something like
```ruby
class PreviewMock
  def initialize(name = '')
    @name = name
  end

  def method_missing(name)
    PreviewMock.new(@name + '.' + name)
  end

  def to_s
    @name
  end
end
```

the pass `PreviewMock.new` to the `@object` when we preview the tenplate.

ofedoren commented:
@ares, thanks for the direction. I've actually tried to implement usage of actual models (`Host::Managed`, etc) and the implementation was pretty simple, although I've decided to drop the idea for now since it will require quite changes in `Editor.js` in core to allow to use custom selectors (currently it's tied to hosts only) or to use own `Editor`, but it will just duplicate a lot of code.

Thus, I continued with your idea and it seems to be working pretty fine. I had to break a bit `Jail` for the mock class, but with safe mode enabled all other classes will behave safely. Even if someone will try to `@object.instance_eval { `something dangerous` }` it will fail.

ofedoren commented:
@ares, could you please take a look? :)

ares commented:
I've tested this with safemode enabled and found a serious problem :-) `@object.exec('ls /')` kills the server (normal exec behavior) and lists all files. I assume the problem is `@source.send(name, *args, &block)`. What is the `@source` in this case?

ofedoren commented:
> I've tested this with safemode enabled and found a serious problem :-) `@object.exec('ls /')` kills the server (normal exec behavior) and lists all files. I assume the problem is `@source.send(name, *args, &block)`. What is the `@source` in this case?

Hmm.. Forgot about this possibility.. I'll probably try to revisit dangerous methods on objects and make them disabled.

@source is the variable with the actual class/object which is being "jailed":

https://github.com/svenfuchs/safemode/blob/master/lib/safemode/core_ext.rb#L17

https://github.com/svenfuchs/safemode/blob/master/lib/safemode.rb#L27

https://github.com/svenfuchs/safemode/blob/master/lib/safemode/jail.rb#L3

I guess we need to add a check
```ruby
if dangerous.contains?(name)
  super
else
  @source.send(name, *args, &block)
end
```

The problem is to find all the dangerous methods and to not forget anything...

ares commented:
I think the dangerous method denying is not very secure approach. If we forget one (like send for example :-) we'll have a problem of if Ruby introduces a new alias for some dangerous method. Is there another way to do this? I don't have a good idea.

ofedoren commented:
> I think the dangerous method denying is not very secure approach. If we forget one (like send for example :-) we'll have a problem of if Ruby introduces a new alias for some dangerous method. Is there another way to do this? I don't have a good idea.

Well, I see only two ways then:

 - 1. Allow this feature only for unsafe mode.
 - 2. Disable preview.
 
 Since `Jail` works in the following way: It has constantly defined default allowed methods for most basic classes, also just a few for any other class, dev can re-open local Jail class to add more allowed methods, otherwise it will just fail. So without this workaround I suggest, we can't do much about it to have safe mode supported. The only possible another workaround is to use all our documented `Jail`s (probably we can even use apipie-dsl to gather documented ones) and allow all of them, but only them.
 
 @ares, what do you think?

ares commented:
Well one thing that would help for sure is to avoid send and use public_send. That way at least you don't have access to methods like exec/spawn. I'd even combine that with not using Object as a parent class for PreviewMock as it contains many methods already (such as exec/spawn). Instead PreviewMock should inherit from BasicObject which is minimal Object class. Then I think we're fine even with safemode.

ofedoren commented:
Thanks, @ares, I forgot that it's totally legal to inherit from `BasicObject` :facepalm:. I've removed `.send` and whole `Jail` since it doesn't work well with `BasicObject`s and if the library will try to `.jail` it it will receive the same object without methods.

ares reviewed (approved):
Just one suggestion, but works well and safely. There are other things we could be adding, e.g. better support for

```
<% @object.each { |h| %>
<%= h.name %>
<% } %>
```

but that's way out of the scope for this PR. Good thing is, preview now don't fail and the static structure is pretty previewable.

Great job, let me know if you want to incorporate the suggestion or I should merge as is.

ofedoren commented:
@ares, thanks, updated. I've re-defined `__send__` although it gives `warning: redefining '__send__' may cause serious problems` :)

Also, added really naive support for blocks with positive arity only, I don't think we need to support all possible use-cases for it.

lzap commented:
I do not see much value in a preview feature that use a different renderer and a different data (mock). Me as a user I would like to see what a webhook actually sends but since this is an event, we can't show this and I believe we should rather remove this feature than create a mock that is on the edge when it comes to security.

How much work is to simply remove the button from the UI?

lzap commented:
Do we have an example of such template that is useful with this mock?

I requested 2.5 CP for the core patch to be able to remove preview functionality in the meantime: https://github.com/theforeman/foreman/pull/8637

lzap commented on app/lib/foreman_webhooks/renderer/preview_mock.rb:
```diff
@@ -0,0 +1,43 @@
+
+      def instance_exec
+        to_s
+      end
```
Unfortunately, Ruby has a hardcoded warning that cannot be removed: foreman_webhooks/renderer/preview_mock.rb:42: warning: redefining `__send__' may cause serious problems

I tried to use `alias` or `alias_method` without any luck, the warning is still there.

I do not understand why we need to override? It works as expected without redefining `__send__`:

```
[1] pry(main)> object = ForemanWebhooks::Renderer::PreviewMock.new
=> @object.inspect@object.inspect
[2] pry(main)> object.__send__(:instance_exec) { 1 + 1 }
=> "@object"
```

lzap replied:
I think we should not override `__send__` the pure existance of this method is to avoid being overriden. If I understand Marek correctly, this was an idea just to prevent when a new method is added to the BasicObject class. I think we can live with that.

I suggest to remove the override and add a unit test that will make sure that Ruby BasicObject contains just the methods we expect.

```
[12] pry(main)> BasicObject.instance_methods
=> [:__send__, :!, :==, :!=, :equal?, :__id__, :__binding__, :instance_eval, :instance_exec]
```

lzap commented:
Found a blocker (?) issue tho: I am able to execute this code when `__send__` is undefined:

```
[11] pry(main)> object.__send__(:__binding__).methods.send(:eval, "1+1")
=> 2
```

I think `__binding__` might be defined by Ruby debugger or pry, that might or might not be available in the production environment.

I am going to put on the record that I don't want this, I am not sure about the value here.

lzap commented:
I can re-review on Wednesday, cheers.

ofedoren commented:
Hmm... Weird, I've added such test just in case in locally it passes...

lzap commented:
There is still the warning which cannot be turned off because of `__send__`.

I suggest the following: let's create a new PR to disable preview so this can go into stable release and we can finish this later.

lzap commented:
Thanks, @ares whats your opinion about that warning? I cannot find a way to get rid of it, will it be safe without redefining `__send__`?

adamruzicka commented:
@ares ping

ofedoren commented:
Since this PR is hanging here for a while, I'll try to make a different approach after I come back :)

ofedoren commented:
Closing in favor of a better approach.
